firmware/command.cpp

```diff
@@ -1,156 +1,548 @@
+      break;
+    case 21:
+      c->cmd_id = 21;
+      break;
+    case 28:
+      c->cmd_id = 28;
+      return; // Early return, G28 ignores parameters
+    case 30:
+      c->cmd_id = 30;
+      return; // Early return, G30 ignores parameters
+    case 90:
+      c->cmd_id = 90;// G90 - Absolute positioning
+      break;
+    case 91:
+      c->cmd_id = 91; // G91 - Relative positioning
+      break;
+    case 92:
+      c->cmd_id = 92; // G92 - Set home Position
+      break;
+    case 93:
+      c->cmd_id = 93; // G93 - Set Table Position
+      break;
+
+    default:
+      Serial.print( F("// warning: unknown command " ) );
+      Serial.print(c->c) ; Serial.print(" ");
+      Serial.println( line );
+      Serial.flush();
+      return; // Early return, unidentified command
+  }
+  // Parse out parameters
+  parseOutParameters(c, buf_ptr);
+  return;
 }
 
+
+
+void parseOutParameters( Command * c, char **buf_ptr) {
+  // Parse out parameters
+  while ( **buf_ptr ) {
+    skip_whitespace( buf_ptr );
+    switch ( toupper(**buf_ptr) ) {
+      case 'X':
+        (*buf_ptr)++;
+        c->x =  (get_float( buf_ptr ) * 100.0);
+
+        break;
+      case 'Y':
+        (*buf_ptr)++;
+        c->y =  (get_float( buf_ptr ) * 100.0);
+
+        break;
+#ifdef Z_DRIVER_ENABLED
+      case 'Z':
+        (*buf_ptr)++;
+        c->z =  (get_float( buf_ptr ) * 100.0);
+
+        break;
+#endif
+#ifdef E_DRIVER_ENABLED
+      case 'E':
+        (*buf_ptr)++;
+        c->e = get_int( buf_ptr );
+        break;
+#endif
+      case 'F':
+        (*buf_ptr)++;
+        c->f = get_int( buf_ptr );
+        modal_feedrate = c->f;
+        break;
+#ifdef SUPPORTARCMOVES
+      case 'I':
+        (*buf_ptr)++;
+        c->i = (get_float( buf_ptr ) * 100.0);
+        break;
+      case 'J':
+        (*buf_ptr)++;
+        c->j = (get_float( buf_ptr ) * 100.0);
+        break;
+#endif
+      case 'S':
+        (*buf_ptr)++;
+        c->s = get_int( buf_ptr );
+        break;
+      case 'P':
+        (*buf_ptr)++;
+        c->p = get_int( buf_ptr );
+        break;
+
+      default:
+        (*buf_ptr)++;
+        break;
+    }
+  }
+}
+
+
 void print_command( HardwareSerial s, Command * c ) {
-	s.print( "// " );
-	s.print( c->cmd_class ); s.print( c->cmd_id );
-	s.print( " X" ); s.print( c->x );
-	s.print( " Y" ); s.print( c->y );
-	s.print( " E" ); s.print( c->e );
-	s.print( " I" ); s.print( c->i );
-	s.print( " J" ); s.print( c->j );
-	s.print( " F" ); s.println( c->f );
-	s.flush();
+  s.print( "// " );
+  s.print( c->cmd_class ); s.print( c->cmd_id );
+  s.print( " X" ); s.print( c->x );
+  s.print( " Y" ); s.print( c->y );
+#ifdef Z_DRIVER_ENABLED
+  s.print( " Z" ); s.print( c->z );
+#endif
+#ifdef E_DRIVER_ENABLED
+  s.print( " E" ); s.print( c->e );
+#endif
+#ifdef SUPPORTARCMOVES
+  s.print( " I" ); s.print( c->i );
+  s.print( " J" ); s.print( c->j );
+#endif
+  s.print( " F" ); s.println( c->f );
+  s.flush();
 }
 
 uint8_t Command::execute() {
```

Comment:
I understand your request, please keep giving me feedback. I will try to amend the code accordingly. One limitation is that I would try to avoid calls to functions in order to save cpu cycles in arduino.
I will study how to economize the code.